File: validationDashboard/validationDashboard/R/validationDashboard.R
#' <Add Title>
#'
#' <Add Description>
#'
#' @import htmlwidgets
#'
#' @export
validationDashboard <- function(data, report, width = NULL, height = NULL, elementId = NULL) {

  # forward options using x
  x = list(
    data = data,
    report = report
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'validationDashboard',
    x,
    width = width,
    height = height,
    package = 'validationDashboard',
    elementId = elementId
  )
}

#' Shiny bindings for validationDashboard
#'
#' Output and render functions for using validationDashboard within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a validationDashboard
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name validationDashboard-shiny
#'
#' @export
validationDashboardOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'validationDashboard', width, height, package = 'validationDashboard')
}

#' @rdname validationDashboard-shiny
#' @export
renderValidationDashboard <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, validationDashboardOutput, env, quoted = TRUE)
}

File: validationDashboard/validationDashboard/inst/htmlwidgets/validationDashboard.js
HTMLWidgets.widget({

  name: 'validationDashboard',

  type: 'output',

  factory: function(el, width, height) {
    console.log("factory");
    
    return {

      renderValue: function(x) {

        console.log("renderValue: ", "#"+el.id, x.data, x.report);
        
        // Transform R data format (Object of arrays) into javascript data format (array of objects):
        var newData = [];
        var keys = Object.keys(x.data);
        if (keys.length === 0) return;
        var cnt = x.data[keys[0]].length;
        for (var i=0; i<cnt; i++) {
          var o = {};
          for (var key of keys) o[key] = x.data[key][i];
          newData.push(o);
        }
        
        new validationDashboard({container: "#"+el.id, data: newData, report: x.report, idcol: "ID"});

      },

      resize: function(width, height) {
        
        console.log("resize", width, height);

      }

    };
  }
});